File: __OTHER_FRAMEWORKS__/grafana/dashboards-as-code-workshop-main/part-one-python-starter/src/grafana.py
import urllib.parse
import json
import os
import typing
import requests
from grafana_foundation_sdk.models.dashboard import Dashboard
from grafana_foundation_sdk.cog.encoder import JSONEncoder


class Config:
    host: str
    user: str
    password: str

    def __init__(self, host: str = "", user: str = "", password: str = ""):
        self.host = host
        self.user = user
        self.password = password

    @classmethod
    def from_env(cls) -> typing.Self:
        return cls(
            host=os.environ.get("GRAFANA_HOST", "localhost:3000"),
            user=os.environ.get("GRAFANA_USER", "admin"),
            password=os.environ.get("GRAFANA_PASSWORD", "admin"),
        )


class Client:
    config: Config

    def __init__(self, config: Config):
        self.config = config

    def find_or_create_folder(self, name: str) -> str:
        auth = (self.config.user, self.config.password)
        response = requests.get(
            f"http://{self.config.host}/api/search?type=dash-folder&query={urllib.parse.quote_plus(name)}",
            auth=auth,
        )
        if response.status_code != 200:
            raise RuntimeError(
                f"could not fetch folders list: expected 200, got {response.status_code}"
            )

        # The folder exists.
        response_json = response.json()
        if len(response_json) == 1:
            return response_json[0]["uid"]

        # The folder doesn't exist: we create it.
        response = requests.post(
            f"http://{self.config.host}/api/folders",
            auth=auth,
            headers={"Content-Type": "application/json"},
            data=json.dumps({"title": name}),
        )
        if response.status_code != 200:
            raise RuntimeError(
                f"could not create new folder: expected 200, got {response.status_code}"
            )

        return response.json()["uid"]

    def persist_dashboard(self, folder_uid: str, dashboard: Dashboard):
        auth = (self.config.user, self.config.password)
        response = requests.post(
            f"http://{self.config.host}/api/dashboards/db",
            auth=auth,
            headers={"Content-Type": "application/json"},
            data=json.dumps(
                {
                    "dashboard": dashboard,
                    "folderUid": folder_uid,
                    "overwrite": True,
                },
                cls=JSONEncoder,
            ),
        )
        if response.status_code != 200:
            raise RuntimeError(
                f"could not persist dashboard: expected 200, got {response.status_code}"
            )

File: __OTHER_FRAMEWORKS__/grafana/gitsync-example-B-main/foundation/src/grafana.ts
import { DashboardBuilder } from "@grafana/grafana-foundation-sdk/dashboard";

export interface Config {
    host: string;
    user: string;
    password: string;
}

interface folder {
    uid: string;
}

export class Client {
    private readonly config: Config;

    constructor(config: Config) {
        this.config = config;
    }

    public static withConfigFromEnv(env: Record<string, string>): Client {
        return new Client({
            host: env['GRAFANA_HOST'] ?? 'localhost:3000',
            user: env['GRAFANA_USER'] ?? 'admin',
            password: env['GRAFANA_PASSWORD'] ?? 'admin',
        });
    }

    public async findOrCreateFolder(name: string): Promise<string> {
        const searchResponse = await fetch(this.url(`/api/search?type=dash-folder&query=${encodeURIComponent(name)}`), {
            headers: {Authorization: this.authHeader()},
        }).then(response => response.json() as Promise<folder[]>);

        // The folder exists.
        if (searchResponse.length === 1) {
            return searchResponse[0].uid;
        }

        const payload = {title: name};
        const reqInit = {
            method: 'POST',
            body: JSON.stringify(payload),
            headers: {
                Authorization: this.authHeader(),
                'Content-Type': 'application/json',
            },
        };
        const createResponse = await fetch(this.url(`/api/folders`), reqInit)
            .then(response => response.json() as Promise<folder>);

        return createResponse.uid;
    }

    public async persistDashboard(folderUid: string, dashboard: DashboardBuilder): Promise<void> {
        const reqInit = {
            method: 'POST',
            body: JSON.stringify({
                dashboard: dashboard.build(),
                folderUid: folderUid,
                overwrite: true,
            }),
            headers: {
                Authorization: this.authHeader(),
                'Content-Type': 'application/json',
            },
        };
        
        await fetch(this.url(`/api/dashboards/db`), reqInit);
    }

    private url(path: string): string {
        return `http://${this.config.host}${path}`;
    }

    private authHeader(): string {
        return "Basic " + Buffer.from(`${this.config.user}:${this.config.password}`).toString('base64');
    }
}
